airflow/contrib/hooks/sagemaker_hook.py

```diff
@@ -0,0 +1,167 @@
+                         'ml.c5.xlarge', 'ml.c5.2xlarge', 'ml.c5.4xlarge',
+                         'ml.c5.9xlarge', 'ml.c5.18xlarge']
+        self.is_key("TrainingJobName", config)
+        self.is_key("AlgorithmSpecification", config)
+        self.is_key("TrainingImage", config['AlgorithmSpecification'],
+                    dict_name="AlgorithmSpecification")
+        self.is_key("TrainingInputMode", config['AlgorithmSpecification'],
+                    dict_name="AlgorithmSpecification")
+        self.is_key("RoleArn", config)
+        self.is_key("InputDataConfig", config)
+        for channel in config["InputDataConfig"]:
+            self.is_key("ChannelName", channel, dict_name="InputDataConfig")
+            self.is_key("DataSource", channel, dict_name="InputDataConfig")
+            self.is_key("S3DataSource", channel['DataSource'], dict_name="DataSource")
+            self.is_key("S3DataType", channel['DataSource']['S3DataSource'],
+                        dict_name="S3DataSource")
+            self.is_key("S3Uri", channel['DataSource']['S3DataSource'],
+                        dict_name="S3DataSource")
+        self.is_key("OutputDataConfig", config)
+        self.is_key("S3OutputPath", config['OutputDataConfig'],
+                    dict_name="OutputDataConfig")
+        self.is_key("ResourceConfig", config)
+        self.is_key("InstanceType", config['ResourceConfig'], "ResourceConfig")
+        self.is_key("InstanceCount", config['ResourceConfig'], "ResourceConfig")
+        self.is_key("VolumeSizeInGB", config['ResourceConfig'], "ResourceConfig")
+        if self.is_key("VpcConfig", config):
+            self.is_key("SecurityGroupIds", config['VpcConfig'], "VpcConfig")
+            self.is_key("Subnets", config['VpcConfig'], "VpcConfig")
+        self.is_key("StoppingCondition", config)
+        if self.is_key("Tags", config):
+            self.is_key("Key", config['Tags'], "Tags")
+            self.is_key("Value", config['Tags'], "Tags")
+        if not config['AlgorithmSpecification']['TrainingInputMode'] \
+                in ('Pipe', 'File'):
+            raise AirflowException("TrainingInputMode should either be Pipe or File.")
+        if not config['InputDataConfig']['DataSource']['S3DataSource']['S3DataType']\
+                in ('ManifestFile', 'S3Prefix'):
+            raise AirflowException("S3DataType should either be "
+                                   "ManifestFile or S3Prefix.")
+        if not config['ResourceConfig']['InstanceType'] in InstanceTypes:
+            raise AirflowException("Invalid InstanceType.")
+        self.checkt_for_url(config['InputDataConfig']
+                                  ['DataSource']['S3DataSource']['S3Uri'])
+
+    def get_conn(self):
+        self.conn = self.get_client_type('sagemaker')
+        return self.conn
+
+    def list_training_job(self, job_name):
+        sagemaker_conn = self.get_conn()
```

Comment:
For all api methods (list/create/describe training/tuning jobs), do we need to get_conn inside all of them? Does a single shared self.conn work for all of them?

Reply:
I believe this shouldn't matter because for now, one hook object will only call one API in a node. But I don't see any point why an additional conn need to be established each time, so I changed the code, and now it uses self.conn instead.